lib/dialer/internal.go

```diff
@@ -160,3 +160,40 @@ func (a fallbackAddr) Network() string {
 func (a fallbackAddr) String() string {
 	return a.addr
 }
+
+// Sort available addresses, preferring unspecified address.
+func tcpAddrLess(i interface{}, j interface{}) bool {
+	iIsUnspecified := false
+	jIsUnspecified := false
+	iAddr := i.(*net.TCPAddr)
+	jAddr := j.(*net.TCPAddr)
+
+	if host, _, err := net.SplitHostPort(iAddr.String()); err == nil {
+		iIsUnspecified = host == "" || net.ParseIP(host).IsUnspecified()
+	}
+	if host, _, err := net.SplitHostPort(jAddr.String()); err == nil {
+		jIsUnspecified = host == "" || net.ParseIP(host).IsUnspecified()
+	}
+
+	if jIsUnspecified != iIsUnspecified {
+		return len(iAddr.Network()) <= len(jAddr.Network())
+	}
+
+	return iIsUnspecified
+}
```

Comment:
This is wrong, as I know from the quic stuff. Also could be refactored and unified.